filebeat/input/journald/pkg/journalctl/journalctl.go

```diff
@@ -44,119 +45,135 @@ type journalctl struct {
+			canceler: canceller,
+			cmd:      cmd,
+			dataChan: make(chan []byte),
+			logger:   logger,
 		}
-	}()
-
-	// This goroutine reads the stdout from the journalctl process and makes
-	// the data available via the `Next()` method.
-	// If the journalctl process exits for any reason, then its stdout is closed
-	// this goroutine gets an EOF error and exits.
-	go func() {
-		defer jctl.logger.Debug("stdout reader goroutine done")
-		defer close(jctl.dataChan)
-		reader := bufio.NewReader(jctl.stdout)
-		for {
-			data, err := reader.ReadBytes('\n')
-			if err != nil {
-				if !errors.Is(err, io.EOF) {
-					var logError = false
-					var pathError *fs.PathError
-					if errors.As(err, &pathError) {
-						// Because we're reading from the stdout from a process that will
-						// eventually exit, it can happen that when reading we get the
-						// fs.PathError below instead of an io.EOF. This is expected,
-						// it only means the process has exited, its stdout has been
-						// closed and there is nothing else for us to read.
-						// This is expected and does not cause any data loss.
-						// So we log at level debug to have it in our logs if ever needed
-						// while avoiding adding error level logs on user's deployments
-						// for situations that are well handled.
-						if pathError.Op == "read" &&
-							pathError.Path == "|0" &&
-							pathError.Err.Error() == "file already closed" {
-							logger.Debugf("cannot read from journalctl stdout: '%s'", err)
+
+		var err error
+		jctl.stdout, err = cmd.StdoutPipe()
+		if err != nil {
+			return &journalctl{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+		}
+		jctl.stderr, err = cmd.StderrPipe()
+		if err != nil {
+			return &journalctl{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+		}
+
+		// This gorroutune reads the stderr from the journalctl process, if the
+		// process exits for any reason, then its stderr is closed, this goroutine
+		// gets an EOF error and exits
+		go func() {
+			defer jctl.logger.Debug("stderr reader goroutine done")
+			reader := bufio.NewReader(jctl.stderr)
+			for {
+				line, err := reader.ReadString('\n')
+				if err != nil {
+					if !errors.Is(err, io.EOF) {
+						logger.Errorf("cannot read from journalctl stderr: %s", err)
+					}
+					return
+				}
+
+				logger.Errorf("Journalctl wrote to stderr: %s", line)
+			}
+		}()
+
+		// This goroutine reads the stdout from the journalctl process and makes
+		// the data available via the `Next()` method.
+		// If the journalctl process exits for any reason, then its stdout is closed
+		// this goroutine gets an EOF error and exits.
+		go func() {
+			defer jctl.logger.Debug("stdout reader goroutine done")
+			defer close(jctl.dataChan)
+			reader := bufio.NewReader(jctl.stdout)
+			for {
+				data, err := reader.ReadBytes('\n')
+				if err != nil {
+					if !errors.Is(err, io.EOF) {
+						var logError = false
+						var pathError *fs.PathError
+						if errors.As(err, &pathError) {
```

Comment:
Why isn't this also handled for stderr?

Reply:
Because we're not constantly reading from `stderr` like there is no data being streamed from it, so this error never appeared there. So there was never a need to handle it. But I can see why it makes sense to handle this in both streams.

Most of this special case handling is to prevent us to send a empty or truncated event to the reader.

Btw, most of the code in this function [was already there](https://github.com/elastic/beats/blob/e864d84e159aea7d3d77ded05d4e028cd55e15bb/filebeat/input/journald/pkg/journalctl/journalctl.go#L52-L160). This PR mostly changes indentation on this function.